### PR TITLE
Parallelize `probs` with OpenMP

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
`probs` is central in circuit simulation measurements.

**Description of the Change:**
Parallelize `probs` loops using OpenMP.

**Benefits:**
Faster execution with several threads.
The following benchmarks are performed on ISAIC's AMD EPYC-Milan Processor using a several core/threads. The times are obtained averaging the computation of `probs(target)` 5 times for various number of targets. We use the last release implementation as a reference. Since #795 brings some speed-ups even for a single thread, this is why we observe speed-ups > number of threads.

speedup_vs_nthreads

Another view on the data is the strong scaling efficiency. It is almost perfect for 2-4 threads, fairly good for 8 threads and diminishes significantly for 16 threads. 

efficiency_vs_nthreads

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by LuisAlfredoNu** [comment]
I think that it would be nice to add a scaling efficiency plot. What do you think @vincentmr ?

File: pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp [hunk 1]
```diff
@@ -128,10 +131,14 @@ class Measurements final
             Gates::getIndicesAfterExclusion(wires, num_qubits), num_qubits);
         const std::size_t n_probs = PUtil::exp2(n_wires);
         std::vector<PrecisionT> probabilities(n_probs, 0);
+        auto *probs = probabilities.data();
         std::size_t ind_probs = 0;
         for (auto index : all_indices) {
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp:136]: Is it not possible to include this for loop in the pragma?

Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp:136]: Yes, but it is typically somewhat small. Might be good to put on the inside too. Few cases to check here.

Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementsLQubit.hpp:136]: Doing a quick test, switching the loops to make them not "perfectly nested". We get better parallel efficiency but worse performance.
efficiency_vs_nthreads_bad

**Review by LuisAlfredoNu — APPROVED** [review]
The implementation has good scaling and achieves an important speedup.

**Review by AmintorDusko — APPROVED** [review]
Nice work! Thank you for that!

**Pull request merged**